### PR TITLE
Use cmake and scikit-build

### PR DESCRIPTION
**Pull request by sayandip18**

### CODE REVIEW

**Comment by sayandip18** [comment]
```
  Building wheel for hello (PEP 517) ... error
  ERROR: Command errored out with exit status 1:
   command: /home/sayandip/anaconda3/bin/python /home/sayandip/anaconda3/lib/python3.7/site-packages/pip/_vendor/pep517/in_process/_in_process.py build_wheel /tmp/tmp811jpg8a
       cwd: /tmp/pip-req-build-2cta3nvn
  Complete output (198 lines):
  Not searching for unused variables given on the command line.
  CMake Deprecation Warning at CMakeLists.txt:1 (cmake_minimum_required):
    Compatibility with CMake < 2.8.12 will be removed from a future version of
    CMake.
  
    Update the VERSION argument <min> value or use a ...<max> suffix to tell
    CMake that the project does not need compatibility with older versions.
  
  
  -- The C compiler identification is GNU 9.3.0
  -- Detecting C compiler ABI info
  -- Detecting C compiler ABI info - done
  -- Check for working C compiler: /usr/bin/cc - skipped
  -- Detecting C compile features
  -- Detecting C compile features - done
  -- The CXX compiler identification is GNU 9.3.0
  -- Detecting CXX compiler ABI info
  -- Detecting CXX compiler ABI info - done
  -- Check for working CXX compiler: /usr/bin/c++ - skipped
  -- Detecting CXX compile features
  -- Detecting CXX compile features - done
  -- Configuring done
  -- Generating done
  -- Build files have been written to: /tmp/pip-req-build-2cta3nvn/_cmake_test_compile/build
  -- Found PythonInterp: /home/sayandip/anaconda3/bin/python (found version "3.7.6")
  -- Found PythonLibs: /home/sayandip/anaconda3/lib/libpython3.7m.so (found version "3.7.6")
  -- The C compiler identification is GNU 9.3.0
  -- The CXX compiler identification is GNU 9.3.0
  -- Detecting C compiler ABI info
  -- Detecting C compiler ABI info - done
  -- Check for working C compiler: /usr/bin/cc - skipped
  -- Detecting C compile features
  -- Detecting C compile features - done
  -- Detecting CXX compiler ABI info
  -- Detecting CXX compiler ABI info - done
  -- Check for working CXX compiler: /usr/bin/c++ - skipped
  -- Detecting CXX compile features
  -- Detecting CXX compile features - done
  -- Release Build
  -- Shared library
  -- Submodule update
  -- Configuring done
  -- Generating done
  CMake Warning:
    Manually-specified variables were not used by the project:
  
      SKBUILD
  
  
  -- Build files have been written to: /tmp/pip-req-build-2cta3nvn/_skbuild/linux-x86_64-3.7/cmake-build
  [1/106] Building CXX object _deps/taco-build/src/CMakeFiles/taco.dir/cuda.cpp.o
  [2/106] Building CXX object _deps/taco-build/src/CMakeFiles/taco.dir/ir_tags.cpp.o
  [3/106] Building CXX object _deps/taco-build/src/CMakeFiles/taco.dir/error.cpp.o
  [4/106] Building CXX object _deps/taco-build/src/CMakeFiles/taco.dir/taco_tensor_t.cpp.o
  [5/106] Building CXX object _deps/taco-build/src/CMakeFiles/taco.dir/target.cpp.o
  [6/106] Building CXX object _deps/taco-build/src/CMakeFiles/taco.dir/type.cpp.o
  [7/106] Building CXX object _deps/taco-build/src/CMakeFiles/taco.dir/format.cpp.o
  [8/106] Building CXX object _deps/taco-build/src/CMakeFiles/taco.dir/codegen/codegen_c.cpp.o
  [9/106] Building CXX object _deps/taco-build/src/CMakeFiles/taco.dir/codegen/codegen.cpp.o
  [10/106] Building CXX object _deps/taco-build/src/CMakeFiles/taco.dir/codegen/module.cpp.o
  [11/106] Building CXX object _deps/taco-build/src/CMakeFiles/taco.dir/error/error_checks.cpp.o
  [12/106] Building CXX object _deps/taco-build/src/CMakeFiles/taco.dir/tensor.cpp.o
  [13/106] Building CXX object _deps/taco-build/src/CMakeFiles/taco.dir/error/error_messages.cpp.o
  [14/106] Building CXX object _deps/taco-build/src/CMakeFiles/taco.dir/codegen/codegen_cuda.cpp.o
  [15/106] Building CXX object _deps/taco-build/src/CMakeFiles/taco.dir/index_notation/index_notation_nodes.cpp.o
  [16/106] Building CXX object _deps/taco-build/src/CMakeFiles/taco.dir/index_notation/index_notation_nodes_abstract.cpp.o
  [17/106] Building CXX object _deps/taco-build/src/CMakeFiles/taco.dir/index_notation/index_notation_printer.cpp.o
  [18/106] Building CXX object _deps/taco-build/src/CMakeFiles/taco.dir/index_notation/index_notation_visitor.cpp.o
  [19/106] Building CXX object _deps/taco-build/src/CMakeFiles/taco.dir/index_notation/index_notation_rewriter.cpp.o
  [20/106] Building CXX object _deps/taco-build/src/CMakeFiles/taco.dir/index_notation/kernel.cpp.o
  [21/106] Building CXX object _deps/taco-build/src/CMakeFiles/taco.dir/index_notation/intrinsic.cpp.o
  [22/106] Building CXX object _deps/taco-build/src/CMakeFiles/taco.dir/index_notation/schedule.cpp.o
  [23/106] Building CXX object _deps/taco-build/src/CMakeFiles/taco.dir/index_notation/provenance_graph.cpp.o
  [24/106] Building CXX object _deps/taco-build/src/CMakeFiles/taco.dir/index_notation/index_notation.cpp.o
  [25/106] Building CXX object _deps/taco-build/src/CMakeFiles/taco.dir/ir/ir_generators.cpp.o
  ....
  [88/106] Building CXX object _deps/taco-build/test/CMakeFiles/taco-test.dir/tests-scheduling.cpp.o
  [89/106] Building CXX object _deps/taco-build/test/CMakeFiles/taco-test.dir/tests-storage.cpp.o
  [90/106] Building CXX object _deps/taco-build/test/CMakeFiles/taco-test.dir/tests-storage_alloc.cpp.o
  [91/106] Building CXX object _deps/taco-build/test/CMakeFiles/taco-test.dir/tests-tensor.cpp.o
  [92/106] Building CXX object _deps/taco-build/test/CMakeFiles/taco-test.dir/tests-scheduling-eval.cpp.o
  [93/106] Building CXX object _deps/taco-build/test/CMakeFiles/taco-test.dir/tests-transpose.cpp.o
  [94/106] Building CXX object _deps/taco-build/test/CMakeFiles/taco-test.dir/tests-transformation.cpp.o
  [95/106] Building CXX object _deps/taco-build/test/CMakeFiles/taco-test.dir/tests-type.cpp.o
  [96/106] Building CXX object _deps/taco-build/test/CMakeFiles/taco-test.dir/tests-workspaces.cpp.o
  [97/106] Building CXX object _deps/taco-build/test/CMakeFiles/taco-test.dir/tests-tensor_types.cpp.o
  [98/106] Building CXX object _deps/taco-build/tools/CMakeFiles/taco-tool.dir/taco.cpp.o
  FAILED: _deps/taco-build/tools/CMakeFiles/taco-tool.dir/taco.cpp.o
  /usr/bin/c++ -DTACO_LINUX -I_deps/taco-src/include -Iinclude -I_deps/taco-src/src -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wmissing-declarations -Woverloaded-virtual -pedantic-errors -Wno-deprecated -std=c++14 -O3 -DNDEBUG -MD -MT _deps/taco-build/tools/CMakeFiles/taco-tool.dir/taco.cpp.o -MF _deps/taco-build/tools/CMakeFiles/taco-tool.dir/taco.cpp.o.d -o _deps/taco-build/tools/CMakeFiles/taco-tool.dir/taco.cpp.o -c _deps/taco-src/tools/taco.cpp
  _deps/taco-src/tools/taco.cpp:35:10: fatal error: taco/version.h: No such file or directory
     35 | #include "taco/version.h"
        |          ^~~~~~~~~~~~~~~~
  compilation terminated.
  [99/106] Building CXX object _deps/taco-build/test/CMakeFiles/taco-test.dir/tests-windowing.cpp.o
  [100/106] Building CXX object _deps/taco-build/test/CMakeFiles/taco-gtest.dir/gtest/gtest-all.cc.o
  ninja: build stopped: subcommand failed.
    File "/tmp/pip-build-env-vrq0awk_/overlay/lib/python3.7/site-packages/skbuild/setuptools_wrap.py", line 589, in setup
      cmkr.make(make_args, env=env)
    File "/tmp/pip-build-env-vrq0awk_/overlay/lib/python3.7/site-packages/skbuild/cmaker.py", line 507, in make
      os.path.abspath(CMAKE_BUILD_DIR())))


-- Trying "Ninja" generator - success
  --------------------------------------------------------------------------------
  
  Configuring Project
    Working directory:
      /tmp/pip-req-build-2cta3nvn/_skbuild/linux-x86_64-3.7/cmake-build
    Command:
      cmake /tmp/pip-req-build-2cta3nvn -G Ninja -DCMAKE_INSTALL_PREFIX:PATH=/tmp/pip-req-build-2cta3nvn/_skbuild/linux-x86_64-3.7/cmake-install -DPYTHON_EXECUTABLE:FILEPATH=/home/sayandip/anaconda3/bin/python -DPYTHON_VERSION_STRING:STRING=3.7.6 -DPYTHON_INCLUDE_DIR:PATH=/home/sayandip/anaconda3/include/python3.7m -DPYTHON_LIBRARY:FILEPATH=/home/sayandip/anaconda3/lib/libpython3.7m.so -DSKBUILD:BOOL=TRUE -DCMAKE_MODULE_PATH:PATH=/tmp/pip-build-env-vrq0awk_/overlay/lib/python3.7/site-packages/skbuild/resources/cmake -DCMAKE_BUILD_TYPE:STRING=Release
  
  Traceback (most recent call last):
  
  An error occurred while building with CMake.
    Command:
      cmake --build . --target install --config Release --
    Source directory:
      /tmp/pip-req-build-2cta3nvn
    Working directory:
      /tmp/pip-req-build-2cta3nvn/_skbuild/linux-x86_64-3.7/cmake-build
  Please see CMake's output for more information.
  ----------------------------------------
  ERROR: Failed building wheel for hello
Failed to build hello
ERROR: Could not build wheels for hello which use PEP 517 and cannot be installed directly

```

**Comment by sayandip18** [comment]
The PEP 517 issue is appearing even when I run `pip install . --no-binary :all:`

**Comment by hameerabbasi** [comment]
Looks like the issue is in the TACO repository, or it's not meant to be nested/used as a repo. Raise an issue there.

**Comment by sayandip18** [comment]
I just ran using `ExternalProject_Add()` to build the project instead of `FetchContent()` and the taco error is no longer there.

**Comment by hameerabbasi** [comment]
Can you push the latest version of the code?

**Comment by sayandip18** [comment]
I am unable to push changes.
```
To https://github.com/sayandip18/sparse.git
 ! [rejected]        build2 -> build2 (fetch first)
error: failed to push some refs to 'https://github.com/sayandip18/sparse.git'
hint: Updates were rejected because the remote contains work that you do
hint: not have locally. This is usually caused by another repository pushing
hint: to the same ref. You may want to first integrate the remote changes
hint: (e.g., 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.

```
Is this because you changed the base?

**Comment by hameerabbasi** [comment]
You need to do `git pull`

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,21 @@
+cmake_minimum_required(VERSION 3.4.0)
+
+project(sparse)
+
+add_subdirectory(sparse)
```
Comment (hameerabbasi) [inline CMakeLists.txt:5]: The issue is on this line according to the [CI logs](https://github.com/pydata/sparse/pull/487/checks?check_run_id=2821436520#step:4:134).

**Comment by sayandip18** [comment]
Removing skbuild leads to a perfect installation in my system with no error

**Comment by hameerabbasi** [comment]
But it also defeats the purpose of the PR, `cmake` is never really called when installing `sparse` with a local build.

**Pull request closed without merging**